js/ui/editableLabel.js

```
@@ -205,7 +195,41 @@ const EditableLabel = new Lang.Class({
             return;
         }
 
-        this._endEditing();
+        this.setMode(EditableLabelMode.DISPLAY);
         this.emit('label-edit-update', text);
-    }
+    },
+
+    setMode: function(mode) {
+        if (this._labelMode == mode)
+            return;
+
+        switch (mode) {
+        case EditableLabelMode.DISPLAY:
+            this._endEditing();
+            break;
+
+        case EditableLabelMode.HIGHLIGHT:
+            this.add_style_pseudo_class('highlighted');
+            this._grabHelper.grab({ actor: this,
+                                    onUngrab: Lang.bind(this, this._onHighlightUngrab) });
+            break;
+
+        case EditableLabelMode.EDIT:
+            // we need to be in highlight mode before reaching the edit mode
+            if (this._labelMode != EditableLabelMode.HIGHLIGHT)
+                this.setMode(EditableLabelMode.HIGHLIGHT);
+
+            // enter the edit mode on highlight ungrab
+            if (this._grabHelper.grabbed)
+                this._grabHelper.ungrab({ actor: this });
+            this.grab_key_focus();
+            break;
+
+        default:
+            log('Invalid mode for EditableLabel: ' + mode);
```

Should probably throw an error here